### ISSUE TITLE
[BUG] [LINUX] Downloaded mkv files have incorrect track metadata (i.e. wrong language tag)

### ISSUE THREAD

**Issue by DarwinAwardWinner** [OpenSTDL/CrunchyDL#32]
**Is your feature request related to a problem? Please describe.**
Attached is an screenshot of MKVToolNix GUI showing the track metadata on a downloaded file. The language tags on the audio and subtitle tracks are all set to "tyj". This makes it impossible to tell which track is which actual language. 

![image](https://github.com/stratuma/Crunchyroll-Downloader-v4.0/assets/228145/9865843e-ea26-47e2-a875-8fed6421d458)

**Describe the solution you'd like**
The program knows which dubs and subs it is downloading, since that is a configured option, so it should add the corresponding tags to the tracks in the output file.

**Describe alternatives you've considered**
As long as the program is consistent with the order in which the different tracks are encoded in the file, I can write my own script to add the appropriate language tag to each stream. But it would be better if the program could do it on its own.

**Comment by stratuma** [comment]
Are you sure you're downloading with my downloader?
I just tested everything a couple of hours ago and everything was fine

**Comment by DarwinAwardWinner** [comment]
The about page says "Crunchyroll Downloader v1.2.2 Made by Stratum". I just installed it today.

In case it matters, my OS:

```
$ lsb_release -a
LSB Version:	n/a
Distributor ID:	ManjaroLinux
Description:	Manjaro Linux
Release:	24.0.2
Codename:	Wynsdey
```

**Comment by stratuma** [comment]
![grafik](https://github.com/stratuma/Crunchyroll-Downloader-v4.0/assets/166541445/2f8310ea-f570-43f5-88a9-9addbb2b3588)

Hmm, It looks like the problem only exists on linux, I will check that later.

**Comment by DarwinAwardWinner** [comment]
Thanks for looking into it. I appreciate the quick response!

**Comment by stratuma** [comment]
No problem.
![grafik](https://github.com/stratuma/Crunchyroll-Downloader-v4.0/assets/166541445/aaa43dd2-d99a-49a7-8a64-155bff3061c8)

Funny that its all set to tyj, even if its not even a language provided by my tool. Maybe a broken ffmpeg build?

**Comment by stratuma** [comment]
I think its a ffmpeg problem, the language codes are valid on windows, so it can only be ffmpeg causing this mess

**Comment by stratuma** [comment]
Or linux uses a other language iso format

**Comment by stratuma** [comment]
Should be fixed with the newest version.

**Comment by DarwinAwardWinner** [comment]
Unfortunately it doesn't seem to be fixed in version 1.2.3:

```
$ mkvinfo -a FILE.mkv | grep -E "^\|  \+ (Track number|Name|Lang|Codec ID)"
|  + Track number: 1 (track ID for mkvmerge & mkvextract: 0)
|  + Language: und
|  + Codec ID: V_MPEG4/ISO/AVC
|  + Track number: 2 (track ID for mkvmerge & mkvextract: 1)
|  + Name: tmp
|  + Language: tmp
|  + Codec ID: A_AAC
|  + Track number: 3 (track ID for mkvmerge & mkvextract: 2)
|  + Name: tmp
|  + Language: tmp
|  + Codec ID: A_AAC
|  + Track number: 4 (track ID for mkvmerge & mkvextract: 3)
|  + Name: tmp
|  + Language: tmp
|  + Codec ID: S_TEXT/ASS
|  + Track number: 5 (track ID for mkvmerge & mkvextract: 4)
|  + Name: tmp[FORCED]
|  + Language: tmp
|  + Codec ID: S_TEXT/ASS
```

![image](https://github.com/OpenSTDL/CrunchyDL/assets/228145/154dd267-0b5d-4983-ae36-e63b2be5ff03)

Version: "Crunchyroll Downloader v1.2.3 Made by OpenSTDL"

**Comment by stratuma** [comment]
I use linux mint and I it looks good on my part, is it maybe a distro problem (uses another iso format)?

**Comment by Ponkhy** [comment]
I'm using MX Linux (Debian 12 Bookworm) and unfortunately also still experience the same issues with v1.2.3.

**Comment by DarwinAwardWinner** [comment]
I don't think it's an iso format issue. When I manually set the languages of the various tracks to "jpn" and "eng" using `mkvpropedit`, then everything looks ok. And these strings match what you showed above.

**Comment by stratuma** [comment]
I think this needs more investigation.

**Comment by orangeMikeB** [comment]
Maybe I found something (at least for Linux, I did not test it on Windows): 

in the [src/api/routes/service/service.service.ts](https://github.com/OpenSTDL/CrunchyDL/blob/4c54d9939143ac83bf59ea0b952badb2e157a91d/src/api/routes/service/service.service.ts#L1401C9-L1469C10) component when trying to find the locale in the locales array, it splits the path but then only uses the first index. So currently it always ends up with the first directory (tmp, since the default path is /tmp) as the argument for the language/title metadata for audio and subtitles. For example, if you change the temporary directory to something else (like /home/temp) you will end up with "home" as language/title.

```js
a.split('/')[1].split('.aac')[0]
```

Maybe replace the repeated split calls with a function to make sure to always get the filename/last part of the path:
```js
const getFileName = function(path: string, ext: string, delimiter: string): string {
    const arr = path.split(delimiter)

    if (arr.length == 0) {
        return "unkown"
    }

    return arr[arr.length - 1].split(ext)[0]
}
```

and use it like so:
```js
options.push(`-metadata:s:a:${index} language=${locales.find((l) => l.locale === getFileName(a, '.aac', '/'))
                          ? locales.find((l) => l.locale === getFileName(a, '.aac', '/'))?.iso
                          : getFileName(a, '.aac', '/') }` )
```

I am a bit curious why it seems to work on Windows.
